Support output descriptor checksums

tochicool opened the pull request:
Possible approach that might resolve #6

tochicool committed:
Support output descriptor checksums

GambolingPangolin commented on src/Language/Bitcoin/Script/Descriptors/Checksum.hs:
```diff
@@ -0,0 +1,85 @@
+import Data.Maybe (fromMaybe)
+import Data.Text (Text)
+import qualified Data.Text as Text
+import Data.Vector (Vector)
```
How about `Data.Vector.Unboxed` here?

tochicool replied:
Thanks, didn't know about that module.

GambolingPangolin commented on src/Language/Bitcoin/Script/Descriptors/Parser.hs:
```diff
@@ -41,7 +46,16 @@ parseDescriptor :: Network -> Text -> Either String OutputDescriptor
 parseDescriptor net = A.parseOnly $ outputDescriptorParser net
 
 outputDescriptorParser :: Network -> Parser OutputDescriptor
```
How about having a more structured parse result?:

```haskell
data ChecksumStatus = Valid | Invalid | Absent

data ChecksumDescriptor = ChecksumDescriptor 
  { descriptor :: OutputDescriptor
  , checksumStatus :: ChecksumStatus
  }
```

That way the application logic can handle the various possibilities without having to choose between parsing logic that does or does-not require the checksum.

tochicool replied:
Nice idea. I've added a similar type to the one you suggested in the syntax module and added a parser for it.

GambolingPangolin commented on src/Language/Bitcoin/Script/Descriptors/Parser.hs:
```diff
@@ -116,3 +130,21 @@ pathP = go Deriv
+        unless (input `validDescriptorChecksum` Text.pack checksum) $
+            case descriptorChecksum input of
+                Nothing -> fail "could not compute checksum"
+                Just actual -> fail $ "provided checksum '" <> checksum <> "' does not match computed checksum '" <> Text.unpack actual <> "'"
```
nit: long line

GambolingPangolin commented on src/Language/Bitcoin/Script/Descriptors/Checksum.hs:
```diff
@@ -0,0 +1,85 @@
+  Just checkSymbols -> 1 == polymodChecksum (expandChecksum desc <> checkSymbols)
+
+descriptorChecksum :: Text -> Maybe Text
+descriptorChecksum desc = Text.pack <$> sequenceA [checksumCharset `charsetGet` ((checksum `shiftR` (5 * (7 - i))) .&. 31) | i <- [0 .. 7]]
```
nit: long line

GambolingPangolin reviewed (commented):
Thank you very much for the patch!

tochicool committed:
Use unboxed vector

tochicool committed:
Use structured parse result

tochicool committed:
Cleanup

GambolingPangolin commented on src/Language/Bitcoin/Script/Descriptors/Syntax.hs:
```diff
@@ -2,6 +2,8 @@ module Language.Bitcoin.Script.Descriptors.Syntax (
     OutputDescriptor (..),
     ScriptDescriptor (..),
     KeyDescriptor (..),
+    ChecksumDescriptor (..),
```
I think it would make more sense to define `ChecksumDescriptor` and `ChecksumStatus` in the `Parser` module (or maybe in `Checksum`).

GambolingPangolin commented on src/Language/Bitcoin/Script/Descriptors.hs:
```diff
@@ -7,6 +7,8 @@ module Language.Bitcoin.Script.Descriptors (
     outputDescriptorAtIndex,
     ScriptDescriptor (..),
     scriptDescriptorAtIndex,
+    ChecksumDescriptor (..),
```
Looking at the haddock IMO it flows better if we move these to the "Parsing" section, right at the top.

GambolingPangolin commented on src/Language/Bitcoin/Script/Descriptors/Parser.hs:
```diff
@@ -37,6 +41,13 @@ import Language.Bitcoin.Utils (
  )
 
+
+parseChecksumDescriptor :: Network -> Text -> Either String ChecksumDescriptor
```
I am OK with just calling this function `parseChecksumDescriptor`, even though it is a breaking change.  Now that we have your logic for dealing with checksums, I don't think anyone would want to use the old `parseDescriptor`.

GambolingPangolin replied:
Oops, I'm trying to say we just make this `parseDescriptor` and not have a separate checksum-unaware version.

tochicool replied:
Makes sense. Was trying not to break the exported functions until I was sure that's what you had in mind.

tochicool committed:
Redefine parsers in terms of ChecksumDescriptor

Breaks existing checksum-unaware parsers and moves ChecksumDescriptor
under parsing.
